diffFiles/src/diff_files.py

```diff
@@ -3,87 +3,107 @@
+                candidate2 = Position(i1 - 1, i2)
+                node1 = result.get_node(candidate1)
+                node2 = result.get_node(candidate2)
+
+                # find the candidate with the bigger length
+                if node1 is None and node2 is None:
+                    next_pos = None
+                    length = 0
+                elif node2 is not None and (node1 is None or node1.length < node2.length):
+                    next_pos = candidate2
+                    length = node2.length
+                else:
+                    next_pos = candidate1
+                    length = cast(Node, node1).length
+
+            return Node(next_pos, length)
+
+        result = Matrix([])
+        for i1, line1 in enumerate(l1):
+            result.nodes.append([])
+            for i2, line2 in enumerate(l2):
+                result.nodes[i1].append(compute_node(i1, i2, line1, line2))
+        return result
+
+
+def build_subsequence_matrix(l1: list[str], l2: list[str]) -> Matrix:
     # we use the matrix as a graph
     # where each node points to one of their neighbors
-    nodes: list[list[Node]] = []
-
-    def get_node(pos: tuple[int, int]) -> Node | None:
-        return nodes[pos[0]][pos[1]] if pos[0] >= 0 and pos[1] >= 0 else None
-
-    def compute_node(i1: int, i2: int) -> Node:
-        next_pos: tuple[int, int] | None
-        length: int
-
-        if l1[i1] == l2[i2]:
-            next_pos = (i1 - 1, i2 - 1)
-            length = nodes[next_pos[0]][next_pos[1]].length + 1
-        else:
-            candidate1 = (i1, i2 - 1)
-            candidate2 = (i1 - 1, i2)
-            node1 = get_node(candidate1)
-            node2 = get_node(candidate2)
-
-            # find the candidate with the bigger length
-            if node1 is None and node2 is None:
-                next_pos = None
-                length = 0
-            elif node2 is not None and (node1 is None or node1.length < node2.length):
-                next_pos = candidate2
-                length = node2.length
-            else:
-                next_pos = candidate1
-                length = cast(Node, node1).length
-
-        return Node(next_pos, length)
-
-    for i1 in range(len(l1)):
-        nodes.append([])
-        for i2 in range(len(l2)):
-            nodes[i1].append(compute_node(i1, i2))
+    nodes = Matrix.make(l1, l2)
     return nodes
 
 
-def compute_common_indices(nodes: list[list[Node]]) -> list[tuple[int, int]]:
-    commonIndices: list[tuple[int, int]] = []
-    if not nodes:
-        return commonIndices
+def compute_common_indices(nodes: Matrix) -> list[Position]:
+    common_indices: list[Position] = []
+    if not nodes.nodes:
+        return common_indices
 
-    pos: tuple[int, int] = (len(nodes) - 1, len(nodes[0]) - 1)
+    pos = Position(len(nodes.nodes) - 1, len(nodes.nodes[0]) - 1)
 
-    def get_node(pos: tuple[int, int]) -> Node:
-        return nodes[pos[0]][pos[1]]
+    def get_node(p: Position) -> Node:
```

Comment:
yeah, p can have negative values if the file is empty.
for the empty file, the algorithm breaks down. In particular, the matrix can't be walked, because the starting position does not exist.

I did not include that edge case, as it makes the program more complicated.